### DISCUSSION TITLE
There are shadows under the systray icons (with systray patch)

### DISCUSSION THREAD

**Discussion by bakkeby** [bakkeby/dwm-flexipatch#134] [Known Issues]
Most systray patches manage systray icons in a window separate from the bar in dwm, and just allocates free space within the bar or next to it for the systray window.

Shadows are added by the compositor, but the compositor has no way to identify this systray window hence there is no way to exclude shadows being drawn for that specific window.

To fix this you can try one of the following methods:

**1)** add class hints that you can use in the compositor configuration when it comes to excluding shadows. In the `clientmessage` function when creating the systray window add the following code \[[ref](https://github.com/bakkeby/dwm-flexipatch/commit/d85dc0404f5542b4d9800e2db91df7a577b7f6f2)\]:
```c
    XClassHint ch = {"dwmsystray", "dwmsystray"};
    XSetClassHint(dpy, c->win, &ch);
```

It should be noted that shadows on systray icons, at least in some versions of picom, is thought to be one cause of dwm (or X) unexpectedly crashing.

Once the class hint is in place add this to the `shadow-exclude` setting in your picom.conf \[[ref](https://www.reddit.com/r/suckless/comments/hzop00/dwm_how_to_remove_picomcompton_shadow_from/)\].
```tcl
shadow-exclude = [
   ...
    "class_g = 'dwmsystray'",
   ...
];
```

**2)** alternatively add this to the `shadow-exclude` setting in your picom.conf:
```tcl
shadow-exclude = [
   ...
   "_NET_WM_WINDOW_TYPE@:32a *= '_NET_WM_WINDOW_TYPE_DOCK'",
   ...
];
```

**Comment by astrolul** [comment]
For some reason the shadow just re-appears sometimes and I don't know why, like for example I left to go get something and I took about 2 minutes then came back, opened discord and the shadow was under the systray. My dotfiles are in the "newdotfiles" repo on my github incase maybe its something wrong with my picom config? I also have my dwm on my profile.

EDIT: it seems as if it works first time but when i restart discord the shadow comes back.

**Comment by glpzzz** [comment]
The first method is working for me. Add the code snippet in the `dwm.c` file and then the exclude instruction in picom. All good

![image](https://github.com/bakkeby/dwm-flexipatch/assets/13596248/14cf79a6-1576-499e-a9fd-698a7f43d38a)
